### PR TITLE
3x faster, support logical types better, simpler interface

### PR DESCRIPTION
**Pull request by AlexeyRaga**
## Changes

- New way of de-conflicting: de-conflicting is done at the schema level (thanks @dwincort)
- Simplified and re-written encoding and decoding functionality (up to 3x faster on containers)
- Property testing for pretty much everything
- Better logical types support (can now properly disambiguate between different logical types for the same base type, like `UTCTime` micros vs. millis)

### CODE REVIEW

File: src/Data/Avro/Encoding/FromAvro.hs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+-- fromRecord :: Schema -> Either String a
+
+-- | Descrives how to convert a given intermediate 'Value' into a Haskell data type.
+class FromAvro a where
+  fromAvro :: Value -> Either String a
+
+instance FromAvro Int where
+  fromAvro (Int _ x)  = Right (fromIntegral x)
+  fromAvro (Long _ x) = Right (fromIntegral x)
+  fromAvro x          = Left ("Unable decode Int from: " <> show (describeValue x))
```
Comment (dwincort) [inline src/Data/Avro/Encoding/FromAvro.hs]: All  of these error messages should probably have the word "to" in them.
```suggestion
  fromAvro x          = Left ("Unable to decode Int from: " <> show (describeValue x))
```

File: src/Data/Avro/Encoding/FromAvro.hs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+instance FromAvro a => FromAvro (Maybe a) where
+  fromAvro (Union _ _ Null) = Right Nothing
+  fromAvro (Union _ _ v)    = Just <$> fromAvro v
+  fromAvro x                = Left ("Unable to decode Maybe from: " <> show (describeValue x))
+  {-# INLINE fromAvro #-}
+
+instance (FromAvro a, FromAvro b) => FromAvro (Either a b) where
+  fromAvro (Union _ 0 a) = Left <$> fromAvro a
+  fromAvro (Union _ 1 b) = Right <$> fromAvro b
+  fromAvro (Union _ n _) = Left ("Unable to decode union value with a position #" <> show n)
```
Comment (dwincort) [inline src/Data/Avro/Encoding/FromAvro.hs]: I wonder if this error message should be something about decoding `Either` _from_ a union value.

File: src/Data/Avro/Encoding/ToAvro.hs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+      maybe (failField fld) (flip runEncoder (S.fldType fld)) (HashMap.lookup (S.fldName fld) env)
+
+-- | Describes how to encode Haskell data types into Avro bytes
+class ToAvro a where
+  toAvro :: Schema -> a -> Builder
+
+instance ToAvro Int where
+  toAvro (S.Long _) i = encodeRaw @Int64 (fromIntegral i)
+  toAvro (S.Int _) i  = encodeRaw @Int32 (fromIntegral i)
+  toAvro s _          = error ("Unable to encode Int as: " <> show s)
```
Comment (dwincort) [inline src/Data/Avro/Encoding/ToAvro.hs]: Shouldn't you be able to encode an `Int` into a schema that expects a `Double`?  Or is that somewhere else...

Comment (AlexeyRaga) [reply src/Data/Avro/Encoding/ToAvro.hs]: Hmm, you are right! I'll see through this and potentially other missing cases. Thanks!

**Pull request merged**